rhine-terminal/src/FRP/Rhine/Terminal.hs

```diff
@@ -0,0 +1,118 @@
+{- | Wrapper to write @terminal@ applications in Rhine, using concurrency.
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RecordWildCards #-}
+module FRP.Rhine.Terminal where
+
+-- base
+import Prelude hiding (putChar)
+import Control.Concurrent ()
+import Control.Concurrent.MVar ()
+import Data.IORef ()
+import Unsafe.Coerce (unsafeCoerce)
+
+-- time
+import Data.Time.Clock ( getCurrentTime )
+
+-- terminal
+-- terminal
+import System.Terminal ( awaitEvent, runTerminalT, Event, Interrupt, TerminalT, MonadInput )
+import System.Terminal.Internal ( Terminal )
+
+-- transformers
+import Control.Monad.Trans.Reader
+
+-- rhine
+import FRP.Rhine.Clock.Proxy ()
+import FRP.Rhine
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Catch (MonadMask)
+
+-- | A clock that ticks whenever events or interrupts on the terminal arrive.
+data TerminalEventClock = TerminalEventClock
+
+instance (MonadInput m, MonadIO m) => Clock m TerminalEventClock
+  where
+    type Time TerminalEventClock = UTCTime
+    type Tag  TerminalEventClock = Either Interrupt Event
+
+    initClock TerminalEventClock = do
+      initialTime <- liftIO getCurrentTime
+      return
+        ( constM $ do
+            event <- awaitEvent
+            time <- liftIO getCurrentTime
+            return (time, event)
+        , initialTime
+        )
+
+instance GetClockProxy TerminalEventClock
+
+instance Semigroup TerminalEventClock where
+  t <> _ = t
+
+-- | A function wrapping `flow` to use at the top level
+-- in order to run a `Rhine (TerminalT t m) cl ()`
+--
+-- Example:
+--
+-- @
+--
+-- mainRhine :: MonadIO m => Rhine (TerminalT LocalTerminal m) TerminalEventClock () ()
+-- mainRhine = tagS >-> arrMCl (liftIO . print) @@ TerminalEventClock
+--
+-- main :: IO ()
+-- main = withTerminal $ \term -> `flowTerminal` term mainRhine
+--
+-- @
+
```

Comment:
Do haddocks work if there is an empty line here? Can you have a look?

Reply:
They work but add a newline, so I've tightened it up.